### ISSUE TITLE
Cycle checking results in inconsistent calculations

### ISSUE THREAD

**Issue by Morendil** [openfisca/openfisca-core#749]
Minimal repro:

```py
# -*- coding: utf-8 -*-
import openfisca_france
from openfisca_core.simulations import Simulation

tax_benefit_system = openfisca_france.CountryTaxBenefitSystem()
situation = {
    "individus": {
        "demandeur": {
            "date_naissance": {
                "2018-10": "1952-01-05",
                "2018-09": "1952-01-05",
                "2018-08": "1952-01-05",
                "2018-07": "1952-01-05"
            },
            "salaire_imposable": {
                "2018": 1000,
                "2017": 1000,
                "2016": 1000
            }
        }
    },
    "familles": {
        "_": {
            "parents": [
                "demandeur"
            ],
            "enfants": [],
        }
    },
    "foyers_fiscaux": {
        "_": {
            "declarants": [
                "demandeur"
            ],
            "personnes_a_charge": []
        }
    },
    "menages": {
        "_": {
            "personne_de_reference": [
                "demandeur"
            ],
            "enfants": []
        }
    }
}

allMonths = ['2018-09']

calculs_ok = [
{'variable': 'rfr', 'periods': ['2016']},
{'variable': 'ass', 'periods': allMonths}
]

calculs_ko = [
{'variable': 'ass', 'periods': allMonths},
{'variable': 'rfr', 'periods': ['2016']}
]

simulation_actuelle = Simulation(
tax_benefit_system=tax_benefit_system,
simulation_json=situation)

for computation in calculs_ok:
    for period in computation['periods']:
        data = simulation_actuelle.calculate(computation['variable'], period)
        print (computation['variable'])
        print (data)

simulation_actuelle = Simulation(
tax_benefit_system=tax_benefit_system,
simulation_json=situation)

for computation in calculs_ko:
    for period in computation['periods']:
        data = simulation_actuelle.calculate(computation['variable'], period)
        print (computation['variable'])
        print (data)
```

The results of computing `ass` and `rfr` should not depend on the order of evaluation, but they do:

```
rfr
[573.99994]
ass
[0.]
ass
[0.]
rfr
[0.]
```

This is impacting France and indirectly Mes-Aides (preventing an upgrade to the latest version of France in production). The France model is very ramified and cycles abound; on inspection I think the current solution (`max_nb_cycles`) is misguided.

Connected to openfisca/openfisca-france#1211

**Comment by Morendil** [comment]
One thing that makes this hard to investigate is that it manifests in many ways, some of which can be seen on a small scale (as in the test above) and some of which have to be embraced at a larger scale. I've created a gist with the [trace of a calculation](https://gist.github.com/Morendil/ac79752fd138a44db65b9c3fd66b0a91) based on a version of Core that removes cycle checks altogether. This trace is from executing [this Mes Aides test](https://github.com/openfisca/openfisca-france/blob/0637381a0591a8bbe8fa4617377d375a53e1f027/tests/mes-aides.gouv.fr/test_mes_aides_546b1b3b5d4a0b016431db4c.yaml). It's 10K lines long - that's how deep the stack reaches.

One of the things that's going in there is that we calculate `salaire_imposable` 6 times for different periods, to calculate `autonomie_financiere`. If you search for `autonomie_financiere` in the trace it quickly becomes apparent that we compute it at ever increasing depths of stack, on each occasion for a period reaching further and further back in time.

The hunch I'm investigating is that **this isn't about detecting "cycles" at all**. It seems clear for this particular observation that the problem can be also framed as **why are we even attempting to compute `salaire_imposable` in the first place, when we don't have a salaire_de_base in the first place ?** This is a semantic determination, not a purely syntactic one. I, a human, know that it's pointless to try and compute a taxable salary when there is no salary information in the situation we are computing.

Is there a clue we could follow ? It seems like there is:

```
class salaire_de_base(Variable):
    value_type = float
    entity = Individu
    label = u"Salaire de base, en général appelé salaire brut, la 1ère ligne sur la fiche de paie"
    set_input = set_input_divide_by_period
    reference = u'http://www.insee.fr/fr/methodes/default.asp?page=definitions/salaire-mensuel-base-smb.htm'
    definition_period = MONTH
```

Well, it doesn't have a default value! That makes sense, because there's really no "default" base salary in salary computation - but there can be the absence of a salary. And in fact we can short-circuit a bunch of calculations if we could detect this. Unfortunately, the code for variables as written has a notion of a "default default value", and so we don't take advantage of this clue.

However, this doesn't save us yet. I've tried forcing Core to skip the computation if the variable being computed was salaire_imposable - and we still hit a cycle - but a different one, which is a clue that cycles in fact abound in the France model. (A good hint that we should solve this problem in a general way and not piecemeal, I think.)

Here is [the resulting trace](https://gist.github.com/Morendil/b9dfe7ef9158dae4f41f49b7f1395a56). This looks more like a "true" circular definition: we have the eligibility for AAH computed based on a bunch of income sources which include another aid called ASI/ASPA, for which eligibility in turn hinges on a bunch of income sources including AAH itself.

However, it's important to note that it's not a real cycle in the sense that each time around we shift the period of evaluation towards the past. It would be reasonable to compute a situation in which we want to know, for instance, what fluctuations exist over time as a result of being granted aids in the past which causes someone's income to go over a threshold which causes them not to be eligible for this aid.

What doesn't really make sense in the context of the test is this: we are evaluating eligibility for ASI/ASPA at infinite depths **for a situation where its age condition is not fulfilled**. Again we have a semantic understanding that we could short-circuit a computation - this time not because a variable is absent but because a variable is present.

This is [not a new issue](https://github.com/openfisca/openfisca-core/issues/363) and is linked to vectorization of the computations. In short, we've assumed so far that _vector computation necessarily means eager evaluation_. We are evaluating ASI/ASPA for a 38-year-old *on the chance that the array might contain elderly people*. We've tried to improve this with the `max_nb_cycles` parameter, but this has a huge problem: `max_nb_cycles` **only stops a computation when a cycle is detected**, which means that we have been at least once through the loop of unnecessary computations.

This is absurd in the singular case, but it also makes kind of an absurdity, unless I'm very mistaken about the execution behaviour, of our performance story for vector computations. The bulk of the population is under 65, so just how much performance we're throwing away *even in a vector context* to run one loop of superfluous evaluations for everyone, relative to what we'd be throwing away by performing the computation sequentially, is an open question.

The good news is that we don't have to choose between the two evils of sequential computation or vast amounts of unnecessary computation, at least I don't believe we do. There seem to be technical solutions ([lazyarray](https://lazyarray.readthedocs.io/en/latest/), cytoolz), or in the worst case we could syntactically elevate eligibility conditions to a special status, perform them always _before_ computing income bases (or other expensive aggregates), and filter the arrays post-eligibility to perform the income basis computation only on entities that pass the first level of eligibility.

**Comment by Morendil** [comment]
Continuing on from the previous episode…

I've investigated **short-circuiting** as suggested above. One problem that immediately arises is that it's not clear how to change the "shape" of a call to `simulation.calculate()`, given that
- dependencies propagate via the stack, but
- the vector being operated upon is *not* itself passed via the stack but in a Holder.

What I'd initially envisioned was something like the following scenario, using `aah` as an example:
- compute the vector `aah_eligible`, a relatively shallow computation
- count how many items in that vector are True and note the indices
- create a new sub-vector of the vector being operated on, containing these indices
- pass this (smaller) population vector on to the formula for `aah_base_ressources`
- "unpack" this small vector upon return, into a larger vector, filling the indices in the original that correspond to eligible individuals

The problem here is that we do not have a way to (temporarily, in stack-like fashion) change the vector being operated on into another, because this context is not part of the execution stack, so this fails at step 3.

At this point I decided to step back from solving the question "how to short-circuit computations", and instead go back to investigating the original problem report: figuring out why `rfr` is zeroed out by computing `ass` first in the above test.

The reason this happens is two-fold:
- the result of a computation is always put in cache, even if the computation failed because of a cycle being detected
- the exception raised on detecting a cycle is re-raised through as many cycles as required to ensure that the computation that fails is one that specified a `max_nb_cycles`

Instrumenting the code in simulations.py to figure out when the rfr variable was being computed and cached, and how this interacted with cycle detection, resulted in this:

```
...
Formula result for rfr [0.] 2014
Caching rfr [0.] 2014
in cycle with retraite_imposable 2016-01
in cycle with revenu_assimile_pension 2016
in cycle with revenu_assimile_pension_apres_abattements 2016
in cycle with traitements_salaires_pensions_rentes 2016
in cycle with rev_cat_tspr 2016
in cycle with rev_cat 2016
in cycle with rbg 2016
in cycle with rng 2016
in cycle with rni 2016
ultimate cycle error for rfr 2016
Formula result for rfr None 2016
Default for rfr [0.] 2016
Caching rfr [0.] 2016
...
```

We see that there is a cycle error in `retraite_imposable`, and this "bubbles up" the stack for a while. The first time we reach, moving up the stack, a computation that sets `max_nb_cycles` is [this line](https://github.com/openfisca/openfisca-france/blob/38c64335dc21a047fa326adeb184c8586fcbd64b/openfisca_france/model/prelevements_obligatoires/prelevements_sociaux/contributions_sociales/remplacement.py#L31):
`rfr = individu.foyer_fiscal('rfr', period = period.n_2, max_nb_cycles = 1)`

This is why it's `rfr` that gets zeroed out and not some other variable.

On branch [simplify-cycle-detection](https://github.com/openfisca/openfisca-core/commits/simplify-cycle-detection), we can see the effect of removing the second rule above (re-raise the error until we hit a line with a [max_nb_cycles](url)): all of the tests keep passing (except the unit tests that checked this rule of cycle detection). So we *might* have a quick-and-dirty fix for Mes Aides' problem until we figure out a more permanent solution.

@guillett Would you please test your problematic calculations from Mes Aides against this particular branch of core ?

**Comment by maukoquiroga** [comment]
Also note that by forcing every for every single `calculate` call  to have `max_nb_cycles = 0`, we obtain the desired result:

```
rfr
[573.99994]
ass
[0.]
ass
[0.]
rfr
[573.99994]
```

**Comment by maukoquiroga** [comment]
> What I'd initially envisioned was something like the following scenario, using aah as an example:
> 
> compute the vector aah_eligible, a relatively shallow computation
> count how many items in that vector are True and note the indices
> create a new sub-vector of the vector being operated on, containing these indices
> pass this (smaller) population vector on to the formula for aah_base_ressources
> "unpack" this small vector upon return, into a larger vector, filling the indices in the original that correspond to eligible individuals

You could achieve this by _boolean indexing_ and _fancy indexing_, however memory performance should be tested as both create copies of the arrays, instead of just _views_. An example:

```python
import numpy                                                                                                                                                                                        

montant_aah = numpy.zeros(int(1e5))                                                                                                                                                                                                                                                                                                                                                      
# array([0., 0., 0., ..., 0., 0., 0.])

aah_eligible = numpy.random.choice(numpy.array([True, False]), int(1e5))                                                                                                                                                                                                                                                                                                                
# array([ True,  True,  True, ...,  True, False,  True])

ahh_eligible_indexes = numpy.where(aah_eligible)[0]                                                                                                                                                 
# array([    0,     1,     2, ..., 99994, 99997, 99999])

# let's say this is the formula for now
aah_base_resource_formula = 100                                                                                                                                                                     

montant_aah[ahh_eligible_indexes] += aah_base_resource_formula                                                                                                                                      
# array([100., 100., 100., ..., 100., 100., 100.])

montant_aah                                                                                                                                                                                        
# array([100., 100., 100., ..., 100.,   0., 100.])
```

**Comment by Morendil** [comment]
Linking to this old discussion on short-circuiting: #363.

**Comment by benjello** [comment]
I am not sure I understand everything but I can give you some input of how we got there for france.
 - We had a simple model which worked yearly and was then made more complex to take into account all the possible time behavior.
 - The model was also made more complex by the introduction of many new variables to take care of new input variables.
 - To alleviate the need to specify every variable we use default for variable without formula and allowed to neutralize some variable so they are no more calculated and return their default value
 - Still the were improper initial states (a set of initial variables) that were resulting in infinite spirals toward the past. They could have been explicitly resolved by expanding the initial set.
 - This solution was rejected because it made the specification of tests too cumbersome and the `max_nb_cycles` parameter. 
 - I never thought that it was a good solution but it was a quick fix.
 - My preferred solution would to stop the spirals toward the past is the following:
  - allow the user be able to specify a backstop date before which every a variable could be reset to its default value or a specified value or as it last non-default value. This could be set variable by variable.      
  - this should not set in the country package but provided by the user when initializing its simulation. 
  - the main idea is provide enough information or allow specific explicit imputation

**Comment by Morendil** [comment]
> provided by the user when initializing its simulation.

Well, if we're looking at what the user provides, we can also look at all the inputs; after all if we can make a decision on whether a formula is calculable or not based on that data, we shouldn't have the user do any work. If we _can't_ make that decision, then it's hard to tell the user how to make it. What advice would we give the user ?
- try the oldest variable in your input (but we could do that for them)
- something else ? but any date more recent than the oldest input could cut off some useful data…
- use trial and error (how can we guarantee that _any_ date will work though?)

> allow the user be able to specify a backstop date before which every a variable could be reset to its default value or a specified value

The first option ("try the oldest variable") is what I experimented my initial investigations after discussing the problem with @guillett and it didn't seem to solve the problem very effectively. In some cases the dependencies are so deep that a few cycles are enough to overflow the stack. So if you have a wide range of dates, you're stuck with the problem.

Admittedly, I did this for all variables and not on a variable-by-variable basis. On the other hand if you have to do this for all variables, it could get very tedious to do it by hand…

I agree with the logic that we should keep the formulas in the country package agnostic of how they're going to be computed, and rely on the inputs. But I think we should be able to "follow the values back in time" automatically, not force the user to do it; and I think we should give strong guarantees that computations always terminate, without ever pushing on the user the responsibility for termination (which is what `max_nb_cycles` also does). We should be able to raise an error statically if a computation is going to overflow the stack.

**Comment by fpagnoux** [comment]
Also, FYI, there is already something in Core to print the trace of a calculation:

```py
simulation = Simulation(tax_benefit_system, simulation_dict, trace = True)
simulation.calculate('disposable_income', '2017-01')
simulation.tracer.print_computation_log()
```

will print

```
  disposable_income<2017-01> >> [3920. 2675.]
    salary<2017-01> >> [4000. 2500.]
    basic_income<2017-01> >> [600. 600.]
      age<2017-01> >> [37 33]
        birth<2017-01> >> ['1980-01-01' '1984-01-01']
    income_tax<2017-01> >> [600. 375.]
      salary<2017-01> >> [4000. 2500.]
    social_security_contribution<2017-01> >> [80. 50.]
      salary<2017-01> >> [4000. 2500.]
```

**Comment by Morendil** [comment]
Fixed in #817
